totoify_grafeas/totoifylib.py

```diff
@@ -0,0 +1,266 @@
+
+    if command is None:
+      self.command = []
+    elif isinstance(command, list):
+      self.command = command
+
+    if byproducts is None:
+      self.byproducts = {"custom_values": {}}
+    elif isinstance(byproducts, dict):
+      self.byproducts = byproducts
+
+    if environment is None:
+      self.environment = {"custom_values": {}}
+    elif isinstance(environment, dict):
+      self.environment = environment
+
+  @staticmethod
+  def from_link(link):
+    link.validate()
+
+    grafeas_link = GrafeasLink()
+
+    # materials and products
+    for item in link.materials:
+      grafeas_link.materials.append({"resource_uri": item,
+                                     "hashes": link.materials[item]})
+
+    for item in link.products:
+      grafeas_link.products.append({"resource_uri": item,
+                                    "hashes": link.products[item]})
+
+    grafeas_link.command = link.command
+
+    for key, value in link.byproducts.items():
+      if key == "return-value":
+        grafeas_link.byproducts["custom_values"][key] = str(value)
+      else:
+        grafeas_link.byproducts["custom_values"][key] = value
+
+    for key, value in link.environment.items():
+      grafeas_link.environment["custom_values"][key] = value
+
+    return grafeas_link
+
+  def __repr__(self):
+    return json.dumps({
+        "command": self.command,
+        "materials": self.materials,
+        "products": self.products,
+        "environment": self.environment,
+        "byproducts": self.byproducts
+    })
+
+  def to_dict(self):
+    return {
+        "command": self.command,
+        "materials": self.materials,
+        "products": self.products,
+        "environment": self.environment,
+        "byproducts": self.byproducts
+      }
+
+
+class GrafeasInTotoOccurrence:
+  """Wrapper to hold a Grafeas occurrence complete with in-toto metadata.
+
+  Provides methods to dump occurrence fields into a json string, and to load
+  the occurrence file into the class.
+
+  Attributes:
+    note_name: A string of the name of the note.
+    kind: Set to the string "INTOTO" to indicate the type of occurrence.
+    resource: A dict of resource uri.
+    intoto: A dictionary with key "signatures", a list of signatures consisting
+        of key id and signature hash, and key "signed", the GrafeasLink class.
+  """
+  note_name = None
+  kind = "INTOTO"
+  resource = {
+    "uri": None
+  }
+  intoto = {
+    "signatures":  [],
+    "signed": None
+  }
+
+  def __init__(self, in_toto_link=None, note_name=None, resource_uri=None):
+    """Initalizes intoto "signed" fields with a GrafeasLink."""
+    if in_toto_link:
+      self.intoto["signed"] = GrafeasLink.from_link(in_toto_link.signed)
+      self.intoto["signatures"] = []
+      for signature in in_toto_link.signatures:
+        self.intoto["signatures"].append({"keyid": signature["keyid"],
+                                          "signature": signature["sig"]})
```

Comment:
Minor design nit: For the sake of consistency, I'd probably do the link-to-occurrence conversion in the `from_link` factory and not in the constructor (similar to how you do it in `GrafeasLink`), because this constructor is also used by `load`, which expects an occurrence file that does not require any conversion. Do you see what I mean?

Reply:
Good point! The constructor now merely returns a new instance of `GrafeasInTotoOccurrence`.